compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Dar.hs

```diff
@@ -206,6 +209,34 @@ buildDar service PackageConfigFields {..} ifDir dalfInput = do
                    , Just pkgId
                    )
 
+-- | Takes a list of paths to dars, composite package name and composite package version
+-- Merges together all the dars without usage checks, generates a main package with 0 modules
+-- Generated package uses latest LF version with matching Major version to the dars given, and the current builtin sdk version.
+buildCompositeDar :: [FilePath] -> LF.PackageName -> LF.PackageVersion -> IO (Zip.ZipArchive ())
+buildCompositeDar darPaths name version = do
+  dars <-
+    forM darPaths $ \darPath -> do
+      bs <- BSL.readFile darPath
+      pure $ ZipArchive.toArchive bs
+
+  let darDalfs = fmap (fmap (second3 BSL.toStrict) . either error id . readDalfsWithMeta) dars
+  darLfVersions <- forM darDalfs $ either (fail . DA.Pretty.renderPretty) pure . decodeArchiveLfVersion . snd3 . mainDalf
+
+  let lfVersion =
+        case nubOrd $ LF.versionMajor <$> darLfVersions of
+          -- Note that this will not select dev. We may want to detect if any of the packages are `X.dev` and use this if so (and warn)
+          [mv] -> LF.defaultOrLatestStable mv
+          xs -> error $ "Dars contained multiple different Major LF versions: " <> show (sort xs)        
+      pkgMeta = LF.PackageMetadata name version Nothing
+      pkg = LF.Package { LF.packageLfVersion = lfVersion, LF.packageModules = NM.empty, LF.packageMetadata = Just pkgMeta }
+      (dalf, pkgId) = encodeArchiveAndHash pkg
```

Comment:
when I started reviewing this PR I was very curious what would be the main dalf of a composite `.dar`; it turned out to be a package without any modules. I wonder if anything on daml or canton relies on there being at least one module in a dalf, but I guess we'll find out :)

Reply:
I'm also a bit anxious about this - atm all composite DAR functionality is only in multi-package.yaml, so we can expect the only people who use it will know it's experimental.